### ISSUE TITLE
Shylu: build failure on MacOS

### ISSUE THREAD

**Issue by ccober6** [trilinos/Trilinos#2419]
I did a pull, and got several build errors on my MacOS.  My last pull was sometime last week.  Attached configure and make logs below.   Did a fresh build; got the same errors.

@trilinos/shylu 

## Expectations
Should be able to build on Mac (has until recently).

## Your Environment

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of 

## Additional Information

* Configure log file:  
[configure.log](https://github.com/trilinos/Trilinos/files/1827260/configure.log)

* Build log file:  
[make.log](https://github.com/trilinos/Trilinos/files/1827277/make.log)

**Comment by ndellingwood** [comment]
@ccober6 Would you also post your configure file or line to reproduce, and also the SHA of the branch this occurred with? 

@kyungjoo-kim looking at the errors several of the following type occur:
```
/Users/ccober/myOffice/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_Chol_External.hpp:54:9: error: no matching function for call to 'single'
        Kokkos::single(Kokkos::PerTeam(member), [&]() {
```

If this build is uses Trilinos after SHA dd587659a82f5b3eefc716c5104f0c8bb09812b7 this may likely be from PR #2234

**Comment by ndellingwood** [comment]
@kyungjoo-kim It looks like this strictly a Serial build with Metis disabled:

`Explicitly disabled TPLs on input (by user or by default):  METIS`

Is Metis required with Tacho? I seem to remember some discussion about needing Metis to test Tacho, but maybe it is not required.

**Comment by ndellingwood** [comment]
I see errors in Tacho compiling on my Mac with gcc/5.4.0, Serial backend, using the configure script below, including the errors provided in the make.log above. 

**Script**:
```
export TRILINOS_HOME=/Users/ndellin/Research/trilinos-dev/Trilinos

TPL_PATH=/Users/ndellin/Research/TPL
BLAS_PATH=${TPL_PATH}/BLAS-3.6.0
LAPACK_PATH=${TPL_PATH}/lapack-3.6.1

PARMETIS_PATH=${TPL_PATH}/parmetis-4.0.3/install
METIS_PATH=${TPL_PATH}/metis-5.1.0/install

SCOTCH_PATH=${TPL_PATH}/scotch_6.0.3

export INSTALL_LOCATION=${PWD}

rm -rf CMakeCache.txt CMakeFiles

cmake \
-D CMAKE_INSTALL_PREFIX:PATH="${INSTALL_LOCATION}" \
-D CMAKE_CXX_FLAGS:STRING="-g" \
-D CMAKE_MAKE_PROGRAM="make" \
-D CMAKE_BUILD_TYPE:STRING=RELEASE \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=TRUE \
-D BUILD_SHARED_LIBS:BOOL=OFF \
-D CMAKE_SKIP_RULE_DEPENDENCY:BOOL=ON \
-D Trilinos_VERBOSE_CONFIGURE:BOOL=OFF \
-D Trilinos_ENABLE_INSTALL_CMAKE_CONFIG_FILES:BOOL=OFF \
-D Trilinos_ENABLE_EXPORT_MAKEFILES:BOOL=ON \
-D Trilinos_DEPS_XML_OUTPUT_FILE:FILEPATH="" \
\
-D TPL_ENABLE_MPI:BOOL=ON \
-D CMAKE_CXX_COMPILER:FILEPATH="`which mpicxx`" \
-D CMAKE_C_COMPILER:FILEPATH="`which mpicc`" \
-D CMAKE_Fortran_COMPILER:FILEPATH="`which mpifort`" \
-D TPL_ENABLE_HWLOC:BOOL=OFF \
-D TPL_ENABLE_BLAS:STRING=ON \
    -D BLAS_LIBRARY_DIRS:FILEPATH=${BLAS_PATH} \
    -D BLAS_LIBRARY_NAMES="blas" \
-D TPL_ENABLE_LAPACK:STRING=ON \
    -D LAPACK_LIBRARY_DIRS:FILEPATH=${LAPACK_PATH} \
    -D LAPACK_LIBRARY_NAMES="lapack" \
-D TPL_ENABLE_METIS:BOOL=ON \
  -D METIS_INCLUDE_DIRS:PATH="${METIS_PATH}/include" \
  -D METIS_LIBRARY_DIRS:PATH="${METIS_PATH}/lib" \
  -D METIS_LIBRARY_NAMES:STRING="metis" \
-D TPL_ENABLE_ParMETIS:BOOL=ON \
  -D ParMETIS_INCLUDE_DIRS:PATH="${PARMETIS_PATH}/include" \
  -D ParMETIS_LIBRARY_DIRS:PATH="${PARMETIS_PATH}/lib" \
  -D ParMETIS_LIBRARY_NAMES:STRING="parmetis" \
-D TPL_ENABLE_Scotch:BOOL=OFF \
  -D Scotch_INCLUDE_DIRS:PATH="${SCOTCH_PATH}/include" \
  -D Scotch_LIBRARY_DIRS:PATH="${SCOTCH_PATH}/lib" \
  -D Scotch_LIBRARY_NAMES:STRING="scotch;scotcherr" \
\
-D Trilinos_ENABLE_ALL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EXAMPLES:BOOL=OFF \
-D Trilinos_ENABLE_DEBUG:BOOL=OFF \
-D Trilinos_ENABLE_OpenMP:BOOL=OFF \
-D Trilinos_ENABLE_KokkosCore:BOOL=ON \
-D Trilinos_ENABLE_KokkosContainers:BOOL=ON \
-D Trilinos_ENABLE_KokkosExample:BOOL=OFF \
-D Trilinos_ENABLE_KokkosAlgorithms:BOOL=ON \
-D Kokkos_ENABLE_Serial:BOOL=ON \
-D Kokkos_ENABLE_OpenMP:BOOL=OFF \
  -D Kokkos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_ShyLU_NodeTacho:BOOL=ON \
  -D ShyLU_NodeTacho_ENABLE_TESTS:BOOL=ON \
${TRILINOS_HOME}
```

**Error types**:
```
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/unit-test/Tacho_TestNumeric.hpp:124:145: error: no matching function for call to 'Tacho::Experimental::NumericTools<Kokkos::complex<double>, Kokkos::Serial>::NumericTools(Tacho::Experimental::ordinal_type, Tacho::Experimental::CrsMatrixBase<Kokkos::complex<double>, Kokkos::Serial>::size_type_array&, Tacho::Experimental::CrsMatrixBase<Kokkos::complex<double>, Kokkos::Serial>::ordinal_type_array&, Tacho::Experimental::CrsMatrixBase<Kokkos::complex<double>, Kokkos::Serial>::value_type_array&, Tacho::Experimental::GraphTools_Metis::ordinal_type_array, Tacho::Experimental::GraphTools_Metis::ordinal_type_array, Tacho::Experimental::ordinal_type, Tacho::Experimental::SymbolicTools::ordinal_type_array, Tacho::Experimental::SymbolicTools::size_type_array, Tacho::Experimental::SymbolicTools::ordinal_type_array, Tacho::Experimental::SymbolicTools::size_type_array, Tacho::Experimental::SymbolicTools::ordinal_type_array, Tacho::Experimental::SymbolicTools::ordinal_type_array, Tacho::Experimental::SymbolicTools::ordinal_type_array, Tacho::Experimental::SymbolicTools::size_type_array, Tacho::Experimental::SymbolicTools::ordinal_type_array, Tacho::Experimental::SymbolicTools::ordinal_type_array)'
                                             S.SupernodesTreeParent(), S.SupernodesTreePtr(), S.SupernodesTreeChildren(), S.SupernodesTreeRoots());
                                                                                                                                                 ^
In file included from /Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/unit-test/Tacho_TestNumeric.hpp:26:0,
                 from /Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/unit-test/Tacho_Test.hpp:7,
                 from /Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/unit-test/Tacho_TestSerial_dcomplex.cpp:18:
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp:242:7: note: candidate: Tacho::Experimental::NumericTools<ValueType, ExecSpace>::NumericTools(Tacho::Experimental::ordinal_type, const size_type_array&, const ordinal_type_array&, const ordinal_type_array&, const ordinal_type_array&, Tacho::Experimental::ordinal_type, const ordinal_type_array&, const size_type_array&, const ordinal_type_array&, const size_type_array&, const ordinal_type_array&, const ordinal_type_array&, const ordinal_type_array&, const size_type_array&, const ordinal_type_array&, const ordinal_type_array_host&) [with ValueType = Kokkos::complex<double>; ExecSpace = Kokkos::Serial; Tacho::Experimental::ordinal_type = int; Tacho::Experimental::NumericTools<ValueType, ExecSpace>::size_type_array = Kokkos::View<long unsigned int*, Kokkos::Serial>; Tacho::Experimental::NumericTools<ValueType, ExecSpace>::ordinal_type_array = Kokkos::View<int*, Kokkos::Serial>; Tacho::Experimental::NumericTools<ValueType, ExecSpace>::ordinal_type_array_host = Kokkos::View<int*, Kokkos::Serial>]
       NumericTools(// input matrix A
       ^
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp:242:7: note:   candidate expects 16 arguments, 17 provided
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp:237:7: note: candidate: Tacho::Experimental::NumericTools<ValueType, ExecSpace>::NumericTools(const Tacho::Experimental::NumericTools<ValueType, ExecSpace>&) [with ValueType = Kokkos::complex<double>; ExecSpace = Kokkos::Serial]
       NumericTools(const NumericTools &b) = default;
       ^
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp:237:7: note:   candidate expects 1 argument, 17 provided
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp:231:7: note: candidate: Tacho::Experimental::NumericTools<ValueType, ExecSpace>::NumericTools() [with ValueType = Kokkos::complex<double>; ExecSpace = Kokkos::Serial]
       NumericTools()
       ^
/Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_NumericTools.hpp:231:7: note:   candidate expects 0 arguments, 17 provided

...
 /Users/ndellin/Research/trilinos-dev/Trilinos/packages/shylu/shylu_node/tacho/src/TachoExp_Chol_External.hpp:54:23: error: no matching function for call to 'single(Kokkos::Im     pl::PerTeamValue, Tacho::Experimental::Chol<ArgUplo, Tacho::Experimental::Algo::External>::invoke(SchedType&, MemberType&, const ViewTypeA&) [with SchedType = const int; Memb     erType = const int; ViewTypeA = Kokkos::View<double**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >; ArgUplo = Tacho::Exp     erimental::Uplo::Upper]::<lambda()>)'       Kokkos::single(Kokkos::PerTeam(member), [&]() {
```

**Comment by kyungjoo-kim** [comment]
@ndellingwood @ccober6  

METIS is a required package and cannot be disabled.

**Comment by ndellingwood** [comment]
@kyungjoo-kim thanks for the info, but in the error I posted above I had Metis enabled.

**Comment by kyungjoo-kim** [comment]
@ndellingwood I reproduced the error.

**Comment by ccober6** [comment]
Sorry for the delay, but here is some of the requested info.

build script:
[build.txt](https://github.com/trilinos/Trilinos/files/1830265/build.txt)

SHA that I first encountered this: a2af31cf097d20e615fbc3484ffc2893e1b8621b

**Comment by ccober6** [comment]
Reading through the comments ...  I do have METIS disabled, but have had it that way for a long time (years?).  Should I explicitly enable it, or taking it out and let it use the default?  

Shouldn't I have gotten a configure error?

**Comment by kyungjoo-kim** [comment]
MY guess is that Tacho should not be enabled in the configure stage as it does not have a required TPL. However, somehow it is enabled and luckily it expresses some problems in my code.

**Comment by ndellingwood** [comment]
@ccober6 a temporary workaround may be to set the following:
`-D Trilinos_ENABLE_ShyLU_NodeTacho:BOOL=OFF \`

@kyungjoo-kim thanks for confirming you reproduced. We can look over the Dependencies file in Tacho and CMake logic in Amesos2 to see how Metis disabled slipped through configuration.

**Comment by ndellingwood** [comment]
@kyungjoo-kim These Tacho failures may be related to Amesos2 failures on the dashboard:
https://testing.sandia.gov/cdash/buildSummary.php?buildid=3449837 
https://testing.sandia.gov/cdash/buildSummary.php?buildid=3449847 
https://testing.sandia.gov/cdash/buildSummary.php?buildid=3449921 

**Sample errors**:
```
/scratch/trilinos/workspace/trilinos-folder/Trilinos_generic_nightly@2/SERIAL_Release_gcc_4.9.3__DEV/Trilinos/packages/shylu/shylu_node/tacho/src/Tacho_Solver.hpp:60:24: warning: ‘Tacho::Solver<double, Kokkos::Serial>::_aj’ will be initialized after [-Wreorder]
     ordinal_type_array _aj; ordinal_type_array_host _h_aj;
                        ^
/scratch/trilinos/workspace/trilinos-folder/Trilinos_generic_nightly@2/SERIAL_Release_gcc_4.9.3__DEV/Trilinos/packages/shylu/shylu_node/tacho/src/Tacho_Solver.hpp:59:53: warning:   ‘Tacho::Solver<double, Kokkos::Serial>::size_type_array_host Tacho::Solver<double, Kokkos::Serial>::_h_ap’ [-Wreorder]
     size_type_array    _ap; size_type_array_host    _h_ap;
                                                     ^
/scratch/trilinos/workspace/trilinos-folder/Trilinos_generic_nightly@2/SERIAL_Release_gcc_4.9.3__DEV/Trilinos/packages/shylu/shylu_node/tacho/src/Tacho_Solver.hpp:106:5: warning:   when initialized here [-Wreorder]
     Solver()
     ^
/scratch/trilinos/workspace/trilinos-folder/Trilinos_generic_nightly@2/SERIAL_Release_gcc_4.9.3__DEV/Trilinos/packages/shylu/shylu_node/tacho/src/Tacho_Solver.hpp:63:24: warning: ‘Tacho::Solver<double, Kokkos::Serial>::_peri’ will be initialized after [-Wreorder]
     ordinal_type_array _peri; ordinal_type_array_host _h_peri;
                        ^
/scratch/trilinos/workspace/trilinos-folder/Trilinos_generic_nightly@2/SERIAL_Release_gcc_4.9.3__DEV/Trilinos/packages/shylu/shylu_node/tacho/src/Tacho_Solver.hpp:62:55: warning:   ‘Tacho::Solver<double, Kokkos::Serial>::ordinal_type_array_host Tacho::Solver<double, Kokkos::Serial>::_h_perm’ [-Wreorder]
     ordinal_type_array _perm; ordinal_type_array_host _h_perm;
```

**Comment by ccober6** [comment]
FYI, adding
-D Trilinos_ENABLE_ShyLU_NodeTacho:BOOL=OFF \

allowed my build to complete, and pass my tests.

**Comment by kyungjoo-kim** [comment]
@ndellingwood Dashboard failures are probably not related to this. I belive that nightly tests only test PT packages and Tacho is not PT.

**Comment by ndellingwood** [comment]
@kyungjoo-kim put in PR #2428 with a fix for the issue, waiting on autotester to kick in.

**Comment by kyungjoo-kim** [comment]
I merged the PR #2428 . This will solve the problem. 

@ccober6 If you can compile without disabling tacho, please close this issue.

**Comment by ccober6** [comment]
Ok, I was able to compile without disabling tacho!  Thanks for the help!
